src/nvim/eval.c

```diff
@@ -15432,34 +15440,25 @@ static void do_sort_uniq(typval_T *argvars, typval_T *rettv, bool sort)
       info.item_compare_func_err = false;
       if (info.item_compare_func != NULL
           || info.item_compare_partial != NULL) {
-          item_compare_func_ptr = item_compare2_keeping_zero;
+        item_compare_func_ptr = item_compare2_keeping_zero;
       } else {
-          item_compare_func_ptr = item_compare_keeping_zero;
-      }
-
-      for (li = l->lv_first; li != NULL && li->li_next != NULL; li = li->li_next) {
-        if (item_compare_func_ptr(&li, &li->li_next) == 0) {
-          ptrs[i++].item = li;
-        }
-        if (info.item_compare_func_err) {
-          EMSG(_("E882: Uniq compare function failed"));
-          break;
-        }
+        item_compare_func_ptr = item_compare_keeping_zero;
       }
 
-      if (!info.item_compare_func_err) {
-        while (--i >= 0) {
-          assert(ptrs[i].item->li_next);
-          li = ptrs[i].item->li_next;
-          ptrs[i].item->li_next = li->li_next;
-          if (li->li_next != NULL) {
-            li->li_next->li_prev = ptrs[i].item;
-          } else {
-            l->lv_last = ptrs[i].item;
+      int idx = 0;
+      for (listitem_T *li = TV_LIST_ITEM_NEXT(l, tv_list_first(l))
+           ; li != NULL
+           ; li = TV_LIST_ITEM_NEXT(l, li)) {
+        listitem_T *const prev_li = TV_LIST_ITEM_PREV(l, li);
+        if (item_compare_func_ptr(&prev_li, &li) == 0) {
+          if (info.item_compare_func_err) {
+            EMSG(_("E882: Uniq compare function failed"));
+            break;
           }
-          tv_list_watch_fix(l, li);
-          tv_list_item_free(li);
-          l->lv_len--;
+          tv_list_item_remove(l, li);
+          li = tv_list_find(l, idx);
```

Comment:
Finding the new `li` looks kind of expansive. Can't `prev_li` be reused to set `li`?

Reply:
No, it will actually be invalid after removal. Not invalid *with the current implementation*, but invalid with array-based list implementation that may trigger realloc() at removal. I know that this is suboptimal, but I also know that `uniq()` is almost never used.

Reply:
After searching for uniq() calls in the plugins that happened to be in my ~/.vam I found exactly *one* user of actual uniq() (that is, neosnippet) and a bunch of uniq() implementations based on dictionary (those that do not do sorting), though I do not quite understand WTF author was thinking when writing such code as

```
function! deoplete#util#uniq(list) abort
  let list = map(copy(a:list), '[v:val, v:val]')
  let i = 0
  let seen = {}
  while i < len(list)
    let key = string(list[i][1])
    if has_key(seen, key)
      call remove(list, i)
    else
      let seen[key] = 1
      let i += 1
    endif
  endwhile
  return map(list, 'v:val[0]')
endfunction
```

(it is not deoplete, same thing is seen in some other plugins). (Basically, what I do not understand is what the first and last map()s are for, from my POV you can just copy list over and do remove().)

Reply:
For reference, my ~/.vam contains something like 200 plugins, though most normally not updated as I only installed them to check something.

Reply:
Oh, I remembered that I actually bothered to clone all repositories from https://github.com/vim-scripts once in Jun 18, may search that for uniq(). The following is all what `ag` managed to find there:

```
% ag '^\s*(?!")\S.*(?<![#a-zA-Z0-9.]|\b\w:)\buniq\(' -G '.*\.vim' ../vim-scripts | tee uniqs.lst
../vim-scripts/AmbiCompletion.git/plugin/ambicompletion.vim:151:            call uniq(candidates)
../vim-scripts/AmbiCompletion.git/plugin/ambicompletion.vim:316:        call uniq(cachewords)
../vim-scripts/AmbiCompletion.git/plugin/ambicompletion.vim:355:        call uniq(cachewords)
../vim-scripts/changesPlugin.git/autoload/changes.vim:783:          call uniq(b:diffhl[i])
../vim-scripts/changesPlugin.git/autoload/changes.vim:1283:     let lines = uniq(lines)
../vim-scripts/changesPlugin.git/autoload/changes.vim:1347:         let g:lines=uniq(g:lines)
../vim-scripts/clickable.vim.git/autoload/vital/_clickable_vim.vim:265:    return uniq(a:list)
../vim-scripts/cmake.vim--Alcin.git/autoload/cmake/extension.vim:49:
../vim-scripts/cmake.vim--Alcin.git/autoload/cmake/extension.vim:50:  call uniq(l:signatures)
../vim-scripts/cmake.vim--Alcin.git/autoload/cmake/extension/ninja.vim:47:  call uniq(l:flags)
../vim-scripts/cmake.vim--Alcin.git/autoload/cmake/extension/ninja.vim:117:  call uniq(l:ninja_file_lines)
../vim-scripts/columnmove.git/autoload/vital/_columnmove.vim:265:    return uniq(a:list)
../vim-scripts/ctrlp-locate.git/autoload/vital/_ctrlp_locate.vim:276:    return uniq(a:list)
../vim-scripts/Dutyl.git/autoload/dutyl/util.vim:81:        return uniq(sort(a:list))
../vim-scripts/incsearch.vim.git/autoload/vital/_incsearch.vim:282:    return uniq(a:list)
../vim-scripts/ingo-library.git/autoload/ingo/compat.vim:93:    return uniq(a:list)
../vim-scripts/open-browser.vim.git/autoload/vital/_openbrowser.vim:265:    return uniq(a:list)
../vim-scripts/projectionist.vim.git/autoload/projectionist.vim:29:    return uniq(a:list)
../vim-scripts/tsuquyomi.git/autoload/vital/_tsuquyomi.vim:266:    return uniq(a:list)
../vim-scripts/unite-spell-suggest.git/autoload/mklib/string.vim:142:  return a:0 && a:1 ? uniq(sort(l:parts)) : l:parts
../vim-scripts/vim-gista.git/autoload/vital/_vim_gista.vim:247:    return uniq(a:list)
../vim-scripts/vim-javacomplete2.git/autoload/javacomplete/server.vim:15:    return uniq(a:list)
../vim-scripts/vim-operator-flashy.git/autoload/vital/_operator_flashy.vim:272:    return uniq(a:list)
../vim-scripts/vim-pandoc-bundle.git/autoload/pandoc/command.vim:72:            return filter(uniq(extend(sort(short_opts), sort(long_opts))), 'v:val != "-:"')
../vim-scripts/vim-signjk-motion.git/autoload/vital/_signjk.vim:272:    return uniq(a:list)
../vim-scripts/vim-snoopy.git/autoload/vital/_snoopy.vim:266:    return uniq(a:list)
../vim-scripts/vison.git/autoload/vital/_vison.vim:266:    return uniq(a:list)
```